server/src/main/java/org/elasticsearch/action/admin/indices/rollover/MetadataRolloverService.java

```diff
@@ -221,14 +222,13 @@ private static NameResolution resolveAliasRolloverNames(Metadata metadata, Index
 
     private static NameResolution resolveDataStreamRolloverNames(Metadata metadata, DataStream dataStream, boolean isFailureStoreRollover) {
         final DataStream.DataStreamIndices dataStreamIndices = dataStream.getDataStreamIndices(isFailureStoreRollover);
-        assert dataStreamIndices.getWriteIndex() != null : "Unable to roll over dataStreamIndices with no indices";
+        assert dataStreamIndices.getIndices().isEmpty() == false || dataStreamIndices.isRolloverOnWrite()
+            : "Unable to roll over dataStreamIndices with no indices";
```

Comment:
Yeah I considered that as well. My reasoning for doing it like this was that trying to roll over a data stream with no backing indices shouldn't be possible anyway due to this line in the constructor https://github.com/elastic/elasticsearch/blob/e257a90f8e488e997b35cd75e09e7128fcd5d8ef/server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java#L180-L180
And otherwise I'm sure something somewhere else will throw an exception or trip an assertion.

I could change the `dataStreamIndices.isRolloverOnWrite()` to `(dataStreamInidces.isRolloverOnWrite() && isFailureStoreRollover)`. Would you prefer that?

Reply:
Hm, I do not have a very strong preference but I would say yes, I prefer that.

Let me explain why, I think that it matches assumption we are checking here. So, the code is covered as you said in general but I do think these assertions are also informative not just protective, so the initial assertion could mislead a developer to believing that this applies to both types.

In the end is up to you, it's not a blocking thing for me.